ext/mbstring/mbstring.c

```diff
@@ -2941,6 +2941,144 @@ PHP_FUNCTION(mb_strtolower)
+typedef enum {
+	MB_LTRIM = 1,
+	MB_RTRIM = 2,
+	MB_BOTH_TRIM = 3
+} MB_TRIM_MODE;
+
+static bool is_trim_wchar(uint32_t w, HashTable *ht)
+{
+	return zend_hash_index_exists(ht, w);
+}
+
+static zend_string* trim_each_wchar(zend_string *str, zend_string *what, MB_TRIM_MODE mode, const mbfl_encoding *enc)
+{
+	unsigned char *in = (unsigned char*)ZSTR_VAL(str);
+	unsigned char *what_in = (unsigned char*)ZSTR_VAL(what);
+	uint32_t wchar_buf[128];
+	uint32_t what_wchar_buf[128];
+	size_t in_len = ZSTR_LEN(str);
+	size_t out_len = 0;
+	size_t what_out_len = 0;
+	size_t what_len = ZSTR_LEN(what);
+	unsigned int state = 0;
+	size_t left = 0;
+	size_t right = 0;
+	size_t total_len = 0;
+	bool first = true;
+
+	if (what_len == 0) {
+		return zend_string_copy(str);
+	}
+
+	HashTable what_ht;
+	zval val;
+	ZVAL_EMPTY_STRING(&val);
+	zend_hash_init(&what_ht, what_len, NULL, NULL, false);
+
+	while (what_len) {
+		what_out_len = enc->to_wchar(&what_in, &what_len, what_wchar_buf, 128, &state);
+		ZEND_ASSERT(what_out_len <= 128);
+		for (size_t i = 0; i < what_out_len; i++) {
+			zend_hash_index_add_new(&what_ht, what_wchar_buf[i], &val);
+		}
+	}
+
+	while (in_len) {
+		out_len = enc->to_wchar(&in, &in_len, wchar_buf, 128, &state);
+		ZEND_ASSERT(out_len <= 128);
+		total_len += out_len;
+
+		if (first && (mode & MB_LTRIM)) {
+			for (size_t i = 0; i < out_len; i++) {
+				uint32_t w = wchar_buf[i];
+				if (is_trim_wchar(w, &what_ht)) {
+					left += 1;
+				} else {
+					first = false;
+					break;
+				}
+			}
+		}
+
+		if (mode & MB_RTRIM) {
+			for (size_t j = 0; j < out_len; j++) {
+				uint32_t w = wchar_buf[j];
+				if (is_trim_wchar(w, &what_ht)) {
+					right += 1;
+				} else {
+					right = 0;
+				}
+			}
+		}
+	}
+
+	zend_hash_destroy(&what_ht);
+
+	return mb_get_substr(str, left, total_len - (right + left), enc);
+}
+
+static zend_string* mb_trim_default_chars(zend_string *str, MB_TRIM_MODE mode, const mbfl_encoding *enc)
+{
+	const uint32_t trim_default_chars[] = {
+		0x20, 0x0C, 0x0A, 0x0D, 0x09, 0x0B, 0x00, 0xA0, 0x1680,
+		0x2000, 0x2001, 0x2002, 0x2003, 0x2004, 0x2005, 0x2006, 0x2007,
+		0x2008, 0x2009, 0x200A, 0x2028, 0x2029, 0x202F, 0x205F, 0x3000,
+		0x85, 0x180E
+	};
+	size_t trim_default_chars_length = sizeof(trim_default_chars) / sizeof(uint32_t);
+	mb_convert_buf buf;
+	mb_convert_buf_init(&buf, trim_default_chars_length, MBSTRG(current_filter_illegal_substchar), MBSTRG(current_filter_illegal_mode));
+
+	enc->from_wchar((uint32_t *)trim_default_chars, trim_default_chars_length, &buf, true);
+
+	zend_string *what = mb_convert_buf_result(&buf, enc);
+	zend_string *result = trim_each_wchar(str, what, mode, enc);
+	zend_string_release_ex(what, false);
+	return result;
+}
+
+static void php_do_mb_trim(INTERNAL_FUNCTION_PARAMETERS, MB_TRIM_MODE mode)
+{
+	zend_string *str;
+	zend_string *what = NULL;
+	zend_string *encoding = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(1, 3)
+		Z_PARAM_STR(str)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STR_OR_NULL(what)
```

Comment:
```suggestion
		Z_PARAM_STR(what)
```

The RFC, nor the stubs, doesn't mention this to be nullable. Maybe also use ``characters`` as a variable name?